sympy/solvers/solvers.py

```diff
@@ -2653,22 +2653,59 @@ def _tsolve(eq, sym, **flags):
-                if not sol_base:
-                    return sol_base  # no solutions to remove so return now
-                return list(ordered(set(sol_base) - set(
-                    _solve(lhs.exp, sym, **flags))))
-            elif (rhs is not S.Zero and
-                        lhs.base.is_positive and
-                        lhs.exp.is_real):
-                return _solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags)
-            elif lhs.base == 0 and rhs == 1:
-                return _solve(lhs.exp, sym, **flags)
             else:
-                raise NotImplementedError
+                # sym in lhs.exp.free_symbols
+                if not rhs:
+                    # f(x)**g(x) only has solutions where f(x) == 0 and g(x) != 0 at
+                    # the same place
+                    sol_base = _solve(lhs.base, sym, **flags)
+                    if not sol_base:
+                        return sol_base  # no solutions to remove so return now
+                    return list(ordered(set(sol_base) - set(
+                    _solve(lhs.exp, sym, **flags))))
+                elif lhs.base.is_positive and lhs.exp.is_real:
+                    return _solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags)
+                elif rhs.is_Integer:
+                    if rhs == 1:
+                        if  lhs.base == 0:
+                            return _solve(lhs.exp, sym, **flags)
+                        else:
+                            # 1 = foo**0, 1**foo, (-1)**even
+                            b1 = _solve(lhs.base - 1, sym, **flags)
+                            b_1 = _solve(lhs.base + 1, sym, **flags)
+                            e0 = _solve(lhs.exp, sym, **flags)
+                            e1=[]
+                            rewrite = lhs.rewrite(exp)
+                            if rewrite != lhs:
+                                e1= _solve(rewrite - rhs, sym, **flags)
+                            sol = list(b1) + list(e0) + list(e1)
+                            for b in b_1:
+                                    if eq.subs(sym, b) == 0:
+                                        sol.append(b)
+                            return list(set(ordered(sol)))
+                    else:
+                        sol=[]
+                        from sympy import divisors
+                        x=divisors(rhs)
+                        for i in x:
+                            base_sols = solve(lhs.base - i, sym) + solve(lhs.base + i, sym)
+                            for s in set(base_sols):
+                                if eq.subs(sym, s) == 0:
+                                    sol.append(s)
+                        e1=[]
+                        rewrite = lhs.rewrite(exp)
+                        if rewrite != lhs:
+                            try:
+                                e1= _solve(rewrite - rhs, sym, **flags)
+                            except:
+                                pass
```

Comment:
The lines above appear twice in the patch. I think it would be better to rearrange the logic to remove the duplication.